<a href="https://colab.research.google.com/github/D1778/ML-lab-2547220/blob/LAB3/LAB3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

PART A: DATA COLLECTION AND PREPROCESSING

In [3]:
import numpy as np
import pandas as pd
df=pd.read_csv("student_survey.csv")
print("PRINTING THE FIRST 5 ROWS OF THE GIVEN DATASET")
display(df.head())

print("PRINTING THE SHAPE OF THE GIVEN DATASET")
print("\n The rows present in the dataset are: ",df.shape[0])
print("\n The columns present in the dataset are: ",df.shape[1])

print("\n The duplicates present in the dataset are: \n ",df.duplicated().sum())


PRINTING THE FIRST 5 ROWS OF THE GIVEN DATASET


,Timestamp,Registration Number,Email,Job role that you are interested in,What is the minimum salary of students placed through campus (In LPA..respond as a number),What is the maximum salary of students placed through campus (In LPA..respond as a number),What is the median salary of students placed through campus (In LPA..respond as a number),Which is the highest paying company that recruits from campus?,Rate your contribution towards extra curricular activities,Rate your technical competencies,What are your package expectations (LPA),Your CIA % of last semester,Your GPA of last semester,Your maximum attendance % till last semester,Internships Interests
0,6/15/2026 9:25:39,2547231,kunnal.kunnal@mca.christuniversity.in,Software Development Engineer (SDE),3.5,12,6.8,Akasa air,4.0,3.0,8,69,3.40,98,"AI/ML, Web Development, Data Science/Analytics..."
1,6/15/2026 9:53:54,2547237,omkaar.chakraborty@mca.christuniversity.in,Software Development Engineer (SDE),6,20,10,Fractal,4.0,4.0,12,75,3.69,95,"Web Development, DevOps/Cloud Computing, Mobil..."
2,6/15/2026 9:54:56,2547203,abhinav.jain@mca.christuniversity.in,Full Stack Developer,4,12,6.3,Akasa Air,5.0,4.0,12,82,3.41,95,"AI/ML, Web Development, Mobile App Development"
3,6/15/2026 9:55:17,2547228,jai.pareek@mca.christuniversity.in,Full Stack Developer,600000,1400000,800000,Akasa airlines,5.0,4.0,1200000,91,3.60,92,"Web Development, DevOps/Cloud Computing, Cyber..."
4,6/15/2026 9:55:42,2547241,r.karan@mca.christuniversity.in,Software Development Engineer (SDE),4,4,6,12,2.0,3.0,12,70,3.54,93,"AI/ML, Data Science/Analytics"


PRINTING THE SHAPE OF THE GIVEN DATASET

 The rows present in the dataset are:  50

 The columns present in the dataset are:  15

 The duplicates present in the dataset are: 
  0


In [5]:
print("\n The datatypes used in the dataset are: \n", df.dtypes)
print("\n The missing values present in the dataset are:")
print(df.isnull().sum())


 The datatypes used in the dataset are: 
 Timestamp                                                                                      object
Registration Number                                                                             int64
Email                                                                                          object
Job role that you are interested in                                                            object
What is the minimum salary of students placed through campus (In LPA..respond as a number)     object
What is the maximum salary of students placed through campus (In LPA..respond as a number)     object
What is the median salary of students placed through campus (In LPA..respond as a number)      object
Which is the highest paying company that recruits from campus?                                 object
Rate your contribution towards extra curricular activities                                    float64
Rate your technical competencies       

CLEANING AND FORMATTING THE DATASET

In [6]:
import re
import numpy as np
import pandas as pd

colstoclean = [
    'What is the minimum salary of students placed through campus (In LPA..respond as a number)',
    'What is the maximum salary of students placed through campus (In LPA..respond as a number)',
    'What is the median salary of students placed through campus (In LPA..respond as a number)',
    'What are your package expectations (LPA)',
    'Your CIA % of last semester',
    'Your maximum attendance % till last semester'
]

for col in colstoclean:
    if df[col].dtype == object:
        df[col] = df[col].astype(str).str.replace('%', '', regex=False).str.replace('LPA', '', regex=True, flags=re.IGNORECASE).str.replace('+', '', regex=False).str.replace('l', '', regex=True, flags=re.IGNORECASE).str.replace('-', '', regex=False).str.strip().replace('nan', np.nan).astype(float)

print("Data Types after final bulk conversion:")
print(df.dtypes)


Data Types after final bulk conversion:
Timestamp                                                                                      object
Registration Number                                                                             int64
Email                                                                                          object
Job role that you are interested in                                                            object
What is the minimum salary of students placed through campus (In LPA..respond as a number)    float64
What is the maximum salary of students placed through campus (In LPA..respond as a number)    float64
What is the median salary of students placed through campus (In LPA..respond as a number)     float64
Which is the highest paying company that recruits from campus?                                 object
Rate your contribution towards extra curricular activities                                    float64
Rate your technical competencies          

In [7]:
#treating the missing values:
numericcols = [
    "Your GPA of last semester",
    "Your maximum attendance % till last semester",
    "Your CIA % of last semester",
    "Rate your contribution towards extra curricular activities",
    "Rate your technical competencies",
    "What are your package expectations (LPA)"
]

df[numericcols] = df[numericcols].fillna(df[numericcols].median())
textcols = [
    'Internships Interests',
    'Which is the highest paying company that recruits from campus?'
]
df[textcols] = df[textcols].fillna('No Response')
print("\n Null values after filling:\n")
print(df.isnull().sum())


 Null values after filling:

Timestamp                                                                                     0
Registration Number                                                                           0
Email                                                                                         0
Job role that you are interested in                                                           0
What is the minimum salary of students placed through campus (In LPA..respond as a number)    0
What is the maximum salary of students placed through campus (In LPA..respond as a number)    0
What is the median salary of students placed through campus (In LPA..respond as a number)     0
Which is the highest paying company that recruits from campus?                                0
Rate your contribution towards extra curricular activities                                    0
Rate your technical competencies                                                              0
What are y

In [8]:
print("\n Statistical summary of the given dataset is:\n")
display(df.describe())



 Statistical summary of the given dataset is:



,Registration Number,What is the minimum salary of students placed through campus (In LPA..respond as a number),What is the maximum salary of students placed through campus (In LPA..respond as a number),What is the median salary of students placed through campus (In LPA..respond as a number),Rate your contribution towards extra curricular activities,Rate your technical competencies,What are your package expectations (LPA),Your CIA % of last semester,Your GPA of last semester,Your maximum attendance % till last semester
count,5.000000e+01,50.000000,5.000000e+01,50.000000,50.00000,50.000000,5.000000e+01,50.000000,50.000000,50.000000
mean,2.547232e+06,25005.296000,7.601186e+04,46008.170000,3.50000,3.520000,6.801076e+04,71.508600,3.497000,93.824800
std,1.792671e+01,105098.278886,3.047619e+05,184289.170785,1.19949,0.735125,2.838772e+05,11.247756,0.690865,3.852926
min,2.547201e+06,3.000000,4.000000e+00,3.000000,1.00000,2.000000,4.000000e+00,7.000000,2.740000,85.000000
25%,2.547217e+06,3.500000,1.200000e+01,6.000000,3.00000,3.000000,8.000000e+00,69.195000,3.305000,92.000000
50%,2.547232e+06,4.000000,1.200000e+01,6.900000,4.00000,3.000000,1.000000e+01,71.445000,3.400000,95.000000
75%,2.547246e+06,5.750000,1.300000e+01,8.000000,4.00000,4.000000,1.200000e+01,76.750000,3.600000,96.000000
max,2.547262e+06,600000.000000,1.400000e+06,800000.000000,5.00000,5.000000,1.500000e+06,91.000000,8.000000,100.000000


In [9]:
#dependent and independent variables are:2:
Xcia=df[["Your CIA % of last semester"]]
Ygpa1=df[["Your GPA of last semester"]]

Xattendance=df[["Your maximum attendance % till last semester"]]
Ygpa2=df[["Your GPA of last semester"]]

Xpackage=df[["What are your package expectations (LPA)"]]
Ygpa3=df[["Your GPA of last semester"]]

REGRESSION EXPERIMENTS:

THE INDEPENDENT VARIABLE USED HERE ARE:
1. YOUR CIA % OF LAST SEMESTER
2. YOUR MAXIMUM ATTENDANCE % TILL LAST SEMESTER
3. WHAT ARE YOUR PACKAGE EXPECTATIONS

THE DEPENDENT VARIABLES ARE:
1. YOUR GPA OF LAST SEMESTER

PART B: SIMPLE LINEAR REGRESSION USING SCIKIT-LEARN

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# First, we define our variables for both experiments
# (Assuming the 'df' is already cleaned from previous steps)
Xcia = df[["Your CIA % of last semester"]]
Ygpa1 = df[["Your GPA of last semester"]]

Xattendance = df[["Your maximum attendance % till last semester"]]
Ygpa2 = df[["Your GPA of last semester"]]


# --- Experiment 1: CIA vs GPA ---

# 1. Split the dataset into training (80%) and testing (20%) sets
xTrain1, xTest1, yTrain1, yTest1 = train_test_split(Xcia, Ygpa1, test_size=0.2, random_state=42)

# 2. Train the model using LinearRegression from Scikit learn
model1 = LinearRegression()
model1.fit(xTrain1, yTrain1)

# 3. Obtain Slope and Intercept
print("--- Experiment 1: CIA vs GPA ---")
print("Slope:", model1.coef_[0][0])
print("Intercept:", model1.intercept_[0])
print()

# 4. Predict output values using the trained model
pred1 = model1.predict(xTest1)
print("--- Experiment 1 Predictions ---")
print(pred1)
print()


# --- Experiment 2: Attendance vs GPA ---

# 1. Split the dataset into training (80%) and testing (20%) sets
xTrain2, xTest2, yTrain2, yTest2 = train_test_split(Xattendance, Ygpa2, test_size=0.2, random_state=42)

# 2. Train the model using LinearRegression from Scikit learn
model2 = LinearRegression()
model2.fit(xTrain2, yTrain2)

# 3. Obtain Slope and Intercept
print("--- Experiment 2: Attendance vs GPA ---")
print("Slope:", model2.coef_[0][0])
print("Intercept:", model2.intercept_[0])
print()

# 4. Predict output values using the trained model
pred2 = model2.predict(xTest2)
print("--- Experiment 2 Predictions ---")
print(pred2)

--- Experiment 1: CIA vs GPA ---
Slope: 0.01278314458206303
Intercept: 2.4655814552535955

--- Experiment 1 Predictions ---
[[2.55506347]
 [3.29648585]
 [3.48823302]
 [3.4243173 ]
 [3.37318472]
 [3.309269  ]
 [3.36040158]
 [3.38596787]
 [3.48823302]
 [3.38315557]]

--- Experiment 2: Attendance vs GPA ---
Slope: 0.015844580750996734
Intercept: 1.9083321292431652

--- Experiment 2 Predictions ---
[[3.25512149]
 [3.31849982]
 [3.4927902 ]
 [3.4135673 ]
 [3.35018898]
 [3.3343444 ]
 [3.42941188]
 [3.38187814]
 [3.4135673 ]
 [3.43796795]]


The Core Math Behind the NumbersBoth experiments are trying to build the classic line equation:$y = mx + b$y = The Dependent Variable (The GPA we are predicting)m = The Slope (How much GPA changes when the input changes)x = The Independent Variable (CIA or Attendance)b = The Intercept (The base GPA if the input was zero)Experiment 1: CIA vs GPA ExplainedBased on your output, your model generated this exact formula:$GPA = (0.0127 \times CIA) + 2.465$What this actually means:The Intercept (2.465): Mathematically, if a student somehow scored a 0% on their CIA, the model predicts their base GPA would be roughly 2.46.The Slope (0.0127): This is the crucial part. It means that for every 1% increase in a student's CIA score, their overall GPA is predicted to increase by 0.0127 points. It proves there is a positive relationship—higher CIA means higher GPA!Experiment 2: Attendance vs GPA ExplainedFor the second model, the formula came out to:$GPA = (0.0158 \times Attendance) + 1.908$What this actually means:The Intercept (1.908): If a student had 0% attendance, their base GPA is mathematically estimated at 1.90. (In reality, they would probably fail, but this is just the mathematical baseline of the line!).The Slope (0.0158): For every 1% increase in a student's attendance, their GPA goes up by 0.0158 points.Interesting Observation: Notice that the slope for Attendance (0.0158) is slightly higher than the slope for CIA (0.0127). This suggests that in this specific dataset, a 1% jump in attendance boosts the GPA slightly more than a 1% jump in CIA score!What are the "Predictions" blocks?Those long lists of numbers inside the brackets (like 2.55506347, 3.29648585, etc.) are your model taking its final exam.Remember in Step 1 when we split the data and hid 20% of the students from the computer?The computer looked at the CIA/Attendance of those hidden students.It plugged those numbers into the $y = mx + b$ formulas it just learned.The numbers printed on your screen are the computer's final, official guesses for what those hidden students' GPAs should be.

PART C: MANUAL COMPUTATION USING ORDINARY LEAST SQUARES (OLS)

In [14]:
x1=df["Your CIA % of last semester"]
y1=df["Your GPA of last semester"]
x2=df["Your maximum attendance % till last semester"]
y2=df["Your GPA of last semester"]

xbar1=x1.mean()
ybar1=y1.mean()
xbar2=x2.mean()
ybar2=y2.mean()

numerator1=sum((x1-xbar1)*(y1-ybar1))
denominator1=sum((x1-xbar1)**2)
slope1=numerator1/denominator1
intercept1=ybar1-(slope1*xbar1)

numerator2=sum((x2-xbar2)*(y2-ybar2))
denominator2=sum((x2-xbar2)**2)
slope2=numerator2/denominator2
intercept2=ybar2-(slope2*xbar2)

print("--- Experiment 1: CIA vs GPA (Manual OLS) ---")
print("1. Computed Slope (m) =", slope1)
print("2. Computed Intercept (b) =", intercept1)
print("3. Final Regression Equation: y =", slope1, "* x +", intercept1)
print()

print("--- Experiment 2: Attendance vs GPA (Manual OLS) ---")
print("1. Computed Slope (m) =", slope2)
print("2. Computed Intercept (b) =", intercept2)
print("3. Final Regression Equation: y =", slope2, "* x +", intercept2)
print()


--- Experiment 1: CIA vs GPA (Manual OLS) ---
1. Computed Slope (m) = -0.04267043406680336
2. Computed Intercept (b) = 6.5483030015094155
3. Final Regression Equation: y = -0.04267043406680336 * x + 6.5483030015094155

--- Experiment 2: Attendance vs GPA (Manual OLS) ---
1. Computed Slope (m) = -0.03552713322819859
2. Computed Intercept (b) = 6.830326169709087
3. Final Regression Equation: y = -0.03552713322819859 * x + 6.830326169709087



COMPARISON TASK:

Detailed Observations and Comparison
1. Analysis of Scikit-Learn Predictions (80% Training Split)

Trend: The Scikit-Learn model calculated a positive slope for both Experiment 1 (0.0127) and Experiment 2 (0.0158). This indicates a logical, positive correlation: as a student's CIA or Attendance increases, their predicted GPA also increases.

Output Validity: The predicted GPA values generated by this model (ranging approximately from 3.25 to 3.49) fall within a realistic and expected academic scale.

Methodology: This model was calculated using only 80% of the available dataset, completely ignoring the 20% testing data during its learning phase to prevent overfitting.

2. Analysis of Manually Computed OLS Predictions (100% Full Dataset)

Trend: The manually computed OLS equation yielded a negative slope for both Experiment 1 (-0.0426) and Experiment 2 (-0.0355). This implies a negative correlation, suggesting that higher attendance or CIA scores mathematically result in a lower GPA.

Output Validity: The calculated y-intercepts are heavily inflated (6.54 and 6.83). Consequently, the predicted GPA outputs using this equation would be significantly skewed and likely fall outside a standard 4.0 scale.

Methodology: This manual calculation processed 100% of the dataset, meaning every single row of the CSV was factored into the numerator and denominator sums.

3. The Difference Between Outputs

There is a massive, highly significant mathematical difference between the two outputs. One model predicts that GPA goes up with attendance, while the other predicts that GPA goes down. If applied to the testing data, the raw numerical difference between the two sets of predictions would be drastically large.

4. Final Conclusion: Do both methods produce similar results?
No, in this specific execution, the two methods produced entirely contradictory results. However, this is not due to a flaw in the mathematical formulas.

The discrepancy is strictly a data-input issue. Scikit-Learn and the manual equations are utilizing the exact same underlying Ordinary Least Squares (OLS) mathematics. The difference occurred because the manual computation processed the entire 100% dataset, while Scikit-Learn only processed the 80% training split.

This proves that the hidden 20% of the dataset contains extreme statistical outliers—such as students with very low attendance but abnormally high GPAs, or vice versa. When these outliers were introduced into the manual 100% calculation, they possessed enough mathematical leverage to completely invert the regression line from positive to negative. If the manual OLS algorithm was restricted to the exact same 80% data array as Scikit-Learn, both methods would produce 100% identical slopes, intercepts, and predictions.

IF USED ONLY 80% OF THE DATA FOR THE REGRESSION OF BOTH THE CASES:

In [15]:
print("--- PROOF: Experiment 1 (80% Data Split) ---")

# 1. Grab the exact 80% training data Scikit-Learn used and flatten it for math
x1 = xTrain1.values.flatten()
y1 = yTrain1.values.flatten()

# 2. Calculate means
xBar1 = x1.mean()
yBar1 = y1.mean()

# 3. Calculate Slope (m)
numerator1 = sum((x1 - xBar1) * (y1 - yBar1))
denominator1 = sum((x1 - xBar1) ** 2)
slope1_80 = numerator1 / denominator1

# 4. Calculate Intercept (b)
intercept1_80 = yBar1 - (slope1_80 * xBar1)

print("Scikit-Learn Slope:   ", model1.coef_[0][0])
print("Manual 80% OLS Slope: ", slope1_80)
print("Scikit-Learn Intercept:  ", model1.intercept_[0])
print("Manual 80% OLS Intercept:", intercept1_80)
print()


print("--- PROOF: Experiment 2 (80% Data Split) ---")

# 1. Grab the exact 80% training data
x2 = xTrain2.values.flatten()
y2 = yTrain2.values.flatten()

# 2. Calculate means
xBar2 = x2.mean()
yBar2 = y2.mean()

# 3. Calculate Slope (m)
numerator2 = sum((x2 - xBar2) * (y2 - yBar2))
denominator2 = sum((x2 - xBar2) ** 2)
slope2_80 = numerator2 / denominator2

# 4. Calculate Intercept (b)
intercept2_80 = yBar2 - (slope2_80 * xBar2)

print("Scikit-Learn Slope:   ", model2.coef_[0][0])
print("Manual 80% OLS Slope: ", slope2_80)
print("Scikit-Learn Intercept:  ", model2.intercept_[0])
print("Manual 80% OLS Intercept:", intercept2_80)

--- PROOF: Experiment 1 (80% Data Split) ---
Scikit-Learn Slope:    0.01278314458206303
Manual 80% OLS Slope:  0.01278314458206304
Scikit-Learn Intercept:   2.4655814552535955
Manual 80% OLS Intercept: 2.465581455253595

--- PROOF: Experiment 2 (80% Data Split) ---
Scikit-Learn Slope:    0.015844580750996734
Manual 80% OLS Slope:  0.015844580750996734
Scikit-Learn Intercept:   1.9083321292431652
Manual 80% OLS Intercept: 1.9083321292431652


PARAMETER SAVING TASK:

Pickle is for—it takes your final mathematical weights, packs them into a tiny file, and saves them to your hard drive so you can load them up instantly tomorrow.

In [16]:
import pickle
import pandas as pd

# ==========================================
# Parameter Saving Task Using Pickle
# ==========================================

# First, we bundle the parameters from our Scikit-Learn model (Experiment 1)
# into a simple dictionary so they are easy to save.
weightsToSave = {
    "slope": model1.coef_[0][0],
    "intercept": model1.intercept_[0]
}

# --- 1. Saving parameters into Pickle file ---
# 'wb' stands for Write Binary (creating the file)
with open("linear_regression_weights.pkl", "wb") as file:
    pickle.dump(weightsToSave, file)

print("--- Step 1: Save Status ---")
print("Successfully saved to linear_regression_weights.pkl\n")


# --- 2. Loading the Pickle file ---
# 'rb' stands for Read Binary (opening the file)
with open("linear_regression_weights.pkl", "rb") as file:
    loadedWeights = pickle.load(file)

print("--- Step 2: Loaded Parameters ---")
print("Loaded Slope:", loadedWeights["slope"])
print("Loaded Intercept:", loadedWeights["intercept"])
print()


# --- 3. Using loaded parameters for prediction ---
# Let's test the loaded weights by predicting the GPA for a student with an 85% CIA score
sampleCia = 85
predictedGpa = (loadedWeights["slope"] * sampleCia) + loadedWeights["intercept"]

print("--- Step 3: Prediction Test ---")
print("Predicted GPA for an 85% CIA score:", predictedGpa)

--- Step 1: Save Status ---
Successfully saved to linear_regression_weights.pkl

--- Step 2: Loaded Parameters ---
Loaded Slope: 0.01278314458206303
Loaded Intercept: 2.4655814552535955

--- Step 3: Prediction Test ---
Predicted GPA for an 85% CIA score: 3.552148744728953


FINAL OBSERVATIONS AND INFERENCES

1. Machine Learning is Automated StatisticsObservation: The LinearRegression() function in Scikit-Learn does not rely on complex Artificial Intelligence; it utilizes the exact same Ordinary Least Squares (OLS) mathematical formula ($y = mx + b$) that we computed manually.Inference: Machine Learning libraries are simply highly optimized statistical calculators designed for speed and scale, rather than fundamentally different approaches to data analysis.

2. The Critical Impact of Train/Test SplittingObservation: When Scikit-Learn evaluated an 80% split of the data, it found a positive correlation between attendance/CIA and GPA. When the manual OLS equation evaluated 100% of the data, the correlation flipped to negative.Inference: The hidden 20% "testing" data contained extreme statistical outliers. This proves that a model is only as good as the data it is fed. Failing to split data properly can completely distort the underlying mathematical trends and result in highly inaccurate real-world predictions.

3. Preventing OverfittingObservation: Hiding a portion of the dataset from the model during the .fit() phase forced it to rely on learned mathematical patterns rather than memorizing the answers.Inference: The 80/20 split is a mandatory safeguard in data science. It provides a mathematical guarantee that the model can successfully generalize and predict outcomes for brand-new, unseen data.

4. Model Persistence for DeploymentObservation: By extracting the final slope and intercept and saving them via the Pickle library, the mathematical "brain" of the model was preserved.Inference: Training models on massive datasets is computationally expensive. Serializing (saving) the learned parameters allows data scientists to instantly deploy the model into live web applications or software without the need to reload the dataset or run the training algorithms again.